matchbox_signaling/src/signaling_server/builder.rs

```diff
@@ -0,0 +1,139 @@
+use crate::{
+    signaling_server::{
+        callbacks::{Callback, SharedCallbacks},
+        handlers::{ws_handler, WsUpgradeMeta},
+        NoOpCallouts, NoState,
+    },
+    topologies::{SignalingStateMachine, SignalingTopology},
+    SignalingCallbacks, SignalingServer, SignalingState,
+};
+use axum::{routing::get, Extension, Router};
+use std::{marker::PhantomData, net::SocketAddr};
+use tower_http::{
+    cors::{Any, CorsLayer},
+    trace::{DefaultOnResponse, TraceLayer},
+    LatencyUnit,
+};
+use tracing::Level;
+
+/// Builder for [`SignalingServer`]s.
+///
+/// Begin with [`SignalingServerBuilder::new`] and add parameters before calling
+/// [`SignalingServerBuilder::build`] to produce the desired [`SignalingServer`].
+pub struct SignalingServerBuilder<Topology, Cb = NoOpCallouts, S = NoState>
+where
+    Topology: SignalingTopology<Cb, S>,
+    Cb: SignalingCallbacks,
+    S: SignalingState,
+{
+    /// The socket address to broadcast on
+    pub(crate) socket_addr: SocketAddr,
+
+    /// The router used by the signaling server
+    pub(crate) router: Router,
+
+    /// Shared callouts used by all signaling servers
+    pub(crate) shared_callbacks: SharedCallbacks,
+
+    /// The callbacks used by the signaling server
+    pub(crate) callbacks: Cb,
+
+    /// The state machine that runs a websocket to completion, also where topology is implemented
+    pub(crate) topology: Topology,
+
+    /// Arbitrary state accompanying a server
+    pub(crate) state: PhantomData<S>,
+}
+
+impl<Topology, Cb, S> SignalingServerBuilder<Topology, Cb, S>
+where
+    Topology: SignalingTopology<Cb, S>,
+    Cb: SignalingCallbacks,
+    S: SignalingState,
+{
+    /// Creates a new builder for a [`SignalingServer`].
+    pub fn new(socket_addr: impl Into<SocketAddr>, topology: Topology, state: S) -> Self {
+        Self {
+            socket_addr: socket_addr.into(),
+            router: Router::new()
+                .route("/", get(ws_handler::<Cb, S>))
+                .route("/:path", get(ws_handler::<Cb, S>))
+                .with_state(state),
+            shared_callbacks: SharedCallbacks::default(),
+            callbacks: Cb::default(),
+            topology,
+            state: PhantomData,
+        }
+    }
+
+    /// Modify the router with a mutable closure. This is where one may apply middleware or other
+    /// layers to the Router.
+    pub fn middleware(mut self, mut alter: impl FnMut(&mut Router)) -> Self {
+        alter(&mut self.router);
+        self
+    }
```

Comment:
Is this perhaps too permissive and ergo a possible foot gun? I feel like we probably want it to be an alias to `Router::layer`

Reply:
I'd rather start permissive and if we get issues/people are using it the wrong way then we can scope it down. Open to opinions

Reply:
That's fair, I suppose it's kind of an expert user feature anyway